### PR TITLE
Parameter refactor

### PR DESCRIPTION
**Pull request by tbarchyn**
## Linked issues / PRs

Replaces #7  
Warns loudly about #72 
Addresses #71 
Addresses #59 
Probably related to #55 
Probably kills #34 
Sets the infrastructure for #29 
Sets the infrastructure for #17 

## Overview

This set of improvements is an attempt to create infrastructure for a more pleasurable relationship with parameters. It does not significantly modify parameters themselves, I know there has been a lot of suggestions for revisions to parameter names, or clean up things here and there to simplify, but this PR only focuses on infrastructure. I've put the 'global' parameters that were inline in `ldar_sim_main.py` into a file called `global_parameters.txt`, but have not brought anything else out from the existing program definitions as globals. This work is outstanding, but probably better done separately to this PR - with appropriate mappings set up to maintain a little bit of reverse compatibility.

## Why?

Here are the basic motivations for improvements in the parameter experience:

1. Some parameters are specified in the code - stylistically it is better if the code exists without the parameters. It also cleans up issues with version control, parameters are totally separate from the code, not subject to any licenses or anything and not caught up in git version control.
2. Program definitions had limitations to where they could be located on the file system - this was hard coded in the code - ideally parameter files and outputs can be anywhere.
3. Global parameters were mined from the 'P_ref' program, or whatever the first program definition was - this is OK if you know this, but is error prone. A global level parameters is needed for these simulation wide parameters.
4. Unused parameters existed in program definitions - this is confusing to users as modifying these parameters does nothing.
5. Reverse compatibility of new parameter files is limited. If a new parameter was added to the model, it must be added to all parameter files that have ever existed otherwise the model won't run.
6. Types of input parameters were not fixed. This is OK for just a python model where type inference is handled softly and usually not a problem, but if interacting with json, or databases, fixing types is important, particularly for Soroush. This matters a lot for lists, where they must all be the same type in json or databases, but can exist as a loose collection of objects natively in Python.
7. Some parameters have unique named keys, instead supplied in lists. Parameters in python lists were referenced by index within the list, this is less clear and harder to work with for users as one must remember the index, and must include all values in the list so the referencing works in the program. This is not really fixed, the program just complains about it in this PR.
8. There is a desire to use different file formats or excel spreadsheets - this needs a 'front door' for parameters to map them to whatever the model uses internally, validate them a lot before releasing to the model, the pursuit of the 'parameter mapper'
9. The web app must not write .txt files and execute them - the web app must write json files, or otherwise pass json parameters to the model due to security concerns with the executablity of .txt files interpreted as python code. Not a real issue for desktop use though.
10. Some parameters were modified by the program. This is a style thing that would be good to fix.
11. All the parameters need to be supplied to run the program - this leads to huge parameter files that must be up to date with the model and require significant learning to specify. Running simple simulations can thus be unnecessarily complex.
12. Parameter files had complex rules around formatting - e.g., 'P_ref.txt' must have a name key of P_ref, etc. Anything else does not work and does not provide feedback to the user why it doesn't work. This is an easy error to make. This is still not resolved for .txt parameter files, but not an issue for yaml or json inputs.
13. There was a lot of copying of parameters required to construct a 'treatment / control' simulations with identical control methods. This is a small thing - but it is error prone.

## How do new changes work?

1. All defined parameters in the inputs_template folder are read in to create master default parameter collections. There is a set of defaults for global parameters, a set of defaults for program parameters, and default parameters for each type of method.
2. Parameters are read in one at a time, from multiple parameter files, in either txt, yaml, or json format.
3. First global parameters are sorted out, and types are checked against the defaults. Not the same type? It throws and error, but in the future should probably crash. Supplied parameters are read on top of the defaults - all simulations run with full suite of parameters, always, with only the keys you want to change changed.
4. Second, program parameters are sorted out, setting out the programs in the simulation, same deal: type check and updating to all program parameters.
5. Third, methods are placed into each program, checking types against the different types of methods, type checking, and updating off the defaults.
6. Fourth, the parameter dictionary is written to the output folder, you can call this file as a parameter file and re-run the simulation if you please - but it updates the metadata in a better format.
6. Finally, the full parameter dictionary is released to the model and it proceeds to do what it did before.

## What are the problems with this?

- Complexity: it is complicated and the documentation needs a good study prior to getting started. Even still, it feels complex.
- Opaqueness: relying upon default parameters requires a lot more of an **iron hand** with new features. The defaults must set new features to be 'off' - and turned on as desired. New features and new parameters cannot change old results.
- Defaults: defaults must be rock solid and well thought out because you can run the model easy without ever seeing them.
- New parameters must map to maintain some reverse compatibility - there is a lot of care required changing parameters to maintain reverse compatiblity and without a doubt a test suite is required to automatically test that new features do not break old simulations and appropriately map through parameters. See the `input_mapper_v1.py` for some discussion on the challenges here.

## Can I use the old way of specifying parameters?

Yep. Just supply the `P_ref.txt`, `P_alt.txt`, style of files as arguments or hardcode it in the first few lines of the model. Ideally put your `P_ref.txt`, etc. in a different folder as the ones in `inputs_template` folder are the defaults now and should not be changed unless changing the actual model and defaults it runs with.

## How to test?

Without command line arguments the branch should run simple test simulation, defined by 2 very short parameter files in the `sample_simulations` directory:

First file defines the program to run, called `P_ref`
```
program_name: P_ref
parameter_level: program
version: '2.0'
method_names:
- OGI
```

Second file defines that OGI method referenced in the first file:

```
parameter_level: method
version: '2.0'
OGI:
  type: OGI
```

Every other parameter runs with default values. For more detailed explanation of how this works, refer to the input documentation - I tried to explain how things work with simple examples and how the modularity can be un-caged.

## Notes on parameter mapping

Mapping parameters, or more complex compatibility operations are actually rather complex and a large maintenance burden. I've started on this effort to try to support older parameter files, but refer to input_mapper_v1.py to see discussion and the various types of hooks that must be specified to map different structures of parameters to compliant formats.

## Is this tested?

Yes and no. I've tried to test most things, but I would not be surprised if there were a few gotchas in how this is set up.

### CODE REVIEW

**Review by tarcadius — COMMENTED** [review]
Hi Tom - I spent a bit of time with this. In general it looks amazing - really great work! A couple high-level things (and will review in detail again on the next round after Mo/Keegan stuff is merged):
1. You're right that it's a bit confusing. The part that's most difficult for me to understand is where the parameters come from. I think I understand now that there are program files in the inputs directory like before, and that these get searched for relevant methods according to what's being requested in a different file that's specified as the main argument to the model? So I guess my main question is that if the main argument to the model (right now P_ref.yaml) only specifies OGI, but P_ref has OGI and MGL, will only OGI be run? It just seems like information is being stored in a lot of different places and it's hard to understand what pieces call what other pieces. I presume that P_ref.yaml goes into the inputs directory and searches for P_ref? But P_ref in the inputs directory already has the OGI method contained within it, so why specify it?
2. Similar to above, it seems like all the P_*.txt files have all of the global parameters in them. So does this mean that the global parameter file doesn't matter because it just gets overwritten each time the model is run? Or does global file supersede the program file? It seems to me that all non-method parameters should be in the global file, and that the program file should just list methods. If the program file wants to deviate from the global parameter file, then it only lists those parameters that will change. I feel like the inputs folder should only list global parameters once (and all parameters that aren't method parameters are global parameters). The global parameters could even just be a full program that runs the baseline. Not sure if this makes sense. 
3. When I run the model with the example 3 .txt files in ldar_sim_main I get a bunch of parameter mismatch type errors. Boolean is not integer and so on. Everything appears to run fine though.
4. I absolutely LOVE the output parameter yaml. This is so great.

I think that's my main thing for now - just wondering if this could be a bit more straightforward in terms of understanding what order the files get called and where LDAR-Sim looks first, second, third, and what takes precedence. In my mind the easiest way to do this is to just have a single global parameter file with everything that runs the baseline condition. Then you have a file for each method - you can send these around to your friends. Then you have a third file - which is what you run. The third file contains the barest bones information - simply what methods are used. Those method files get pulled in from method files. All program params come from one global file (unless specified in the program 'run' file. This is maybe what you've already done and I'm just slow!!!!
4. 

Apologies if I'm just missing things and asking bad questions/proposing bad ideas!

File: LDAR_Sim/model_code/input_mapper_v1.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+# ------------------------------------------------------------------------------
+# Program:     The LDAR Simulator (LDAR-Sim)
+# File:        LDAR-Sim input manager
```
[the remark this answers is not in the document]
Comment (tbarchyn) [reply LDAR_Sim/model_code/input_mapper_v1.py:3]: Yep

File: LDAR_Sim/model_code/input_manager.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+import yaml
+import json
+import glob
+import sys
+import time
+
+from input_mapper_v1 import input_mapper_v1
+
+
+def check_types(default, test, omit_keys = None, fatal = False):
```
[the remark this answers is not in the document]
Comment (tbarchyn) [reply LDAR_Sim/model_code/input_manager.py:33]: I'll check PEP 8 - maybe more clear to put this recursive function in separate file.

**Comment by tbarchyn** [comment]
Thanks @tarcadius !

### Re: 1 (where parameters come from)
Yes, I'll expand the documentation to explain this - it wasn't clear enough - thanks for pointing this out!

Global parameters:
defaults (global_parameters.txt) + any supplied inputs -> globals run in model

Program parameters (for each program):
defaults + any supplied inputs -> program parameters run in the model

Method parameters (for each method type):
defaults for each method (mined from 'P_ref.txt') + any supplied inputs -> methods available for use in programs.

Aside on types vs names for methods:
Methods are semantically changed from 'aircraft', 'OGI' whatever, to 'Kairos_aircraft' which has a type of 'aircraft', runs the aircraft module. I think this is a bug in this PR that I'll need to fix revisiting this today - but the concept of having methods with certain names that can be dialed up, modularized, etc. is a change. This change is echoed in changes to method modules, I will add aligning terminology to my to do list.

Defaults
Nothing in the inputs_template folder gets run - these only exist to define the default parameters for each level of the hierarchy that are subsequently built upon.

This is a big change because when you change the P_* files, you are making a change to the defaults - which is a change to how the model runs and would now require a pull request and explanation of what is being changed, etc. It becomes a big deal because it can affect anyone who uses the model and doesn't set that default themselves!

### Re: 2 (location of different parameters within the hierarchy)
This PR sets up the infrastructure - I've shied away from actually changing anything beyond setting up the global parameters that were previously mined from the first program supplied and set as local variables outside the starmap, things like output directory, n_processes, etc.

Other parameters could be globals, or program parameters, or whatever, I've tried to only make bare minimum changes.

My thinking here is all subsequent parameter changes like changing names of parameters, or moving them around from program to changing units, or whatever - all these changes need to be separate PRs that come in with rationale and create the appropriate changes to the input mapper to maintain compatibility. I've tried to set out all the types of changes that need to be considered and really thought about in the default parameter mapper file.

### Re: parameter files changing parameters in other levels of the hierarchy
I've maintained reverse compatibility with version 1 parameter files to do as it did before. Moving forward though, allowing out of hierarchy changes to parameters as a general concept and functionality feels very scary and I'm thinking it should be separate PR that builds on this and sets out the rules for what is allowed, what warns, etc. There is also a lot of thought required for the input mapping and how to manage that.

### Re: 3 (parameter mismatch warnings)
Yeah it warns about lists of parameters with different types
```
'sandwich': ['bread', 3, 'sriracha'],
```
It warns because `bread` is string, and `3` is integer. As part of best practice improvements to parameters that I've tried to synthesize in the documentation, we really should migrate to:
```
'sandwich': {
  'outside': 'bread',
  'n_slices_cheese': 3,
  'sauce': 'sriracha',
},
```
That is basically what it is warning about - trying to coerce the transition to defined parameters. Lists probably should only be used for things that are all the same type and semantically the same, like a list of speeds to sample from or something like that.

However, these 'cleanups' already set up a huge amount of work. If @soroushojagh  sets the sandwich key as a list in the web app side of things, as he probably does right now already, we now must make an entry to the parameter mapper to do something like this:
```
# web_app_parameters are the parameters coming in from the web app from Soroush
'sandwich': {
  'outside': web_app_parameters['sandwich'][0],
  'n_slices_cheese': web_app_parameters['sandwich'][1],
  'sauce': web_app_parameters['sandwich'][2],
}
```
Otherwise parameters coming from Soroush will break the program. There is something in the model like this that will crash if a list is supplied as the sandwich parameter key (as it used to be). The mapper must deal with the change before it gets released to the model.
```
def make_sandwich(parameters):
   put_outside_on_table (type_of_outside = parameters['sandwich']['outside'])
   add_cheese (number_slices = parameters['sandwich']['n_slices'])
   add_sauce (sauce_type = parameters['sandwich']['sauce'])
   eat_sandwich()
```
This is the reverse compatibility trap that I get in a tizzy about! I worry we're already in the trap! On the python end, we are now going to be supporting the specifications that @soroushojagh  has set out until his first version of the web app gets fully replaced with a newer version, and probably even longer until his newest version gets totally stable as I'm certain Soroush may occasionally want to fall back to the version he is making now. I worry we already must support the specification that @soroushojagh  has set up for probably the rest of 2021 - any subsequent changes must already go through the whole rigamarole of setting up the appropriate entry in the parameter mapper, etc.

To do for @tbarchyn:
1. Check overwrites of dictionary children - allow sub-key updates with recursion?
2. Check 'type' argument and relationship with method code - align terminology. Concepts already sound aligned, terminology should be aligned though.
3. Allow transitions between floats and integers with warnings! This is reasonably 'safe' in python, safer from int to float than from float to int.
4. expand documentation.
5. check what happens when nothing runs.
6. check 'type' for methods calls proper code, align with methods branch - they probably have this dialed in.

**Comment by tbarchyn** [comment]
@soroushojagh  - can you please create a branch of the parameters that your parameter database is built upon and tag me directly so I can build a parameter mapper that maps your parameter definitions to the v2 model. I can leave this work to you if you prefer - but please let me know as soon as possible.

File: LDAR_Sim/model_code/input_mapper_v1.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    # old parameter files continue to use the misspelled value
+    if parameters['calculation_method'] == 'categoricle':
+        parameters['calculation_method'] = 'categorical'
+
+    # map the leak rate parameter from one unit to another, the newest version uses a different unit, where 0.003332 is
+    # the conversion factor
+    parameters['leak_rate] = parameters['leak_rate'] * 0.003332
+
+    4) Key-change hooks: these hooks change the key name to a new key name to map the old name to the new name.
+
```
Comment (tbarchyn) [inline LDAR_Sim/model_code/input_mapper_v1.py:55]: @tarcadius @soroushojagh - here is a synopsis of the types of changes that must be specifically dealt with to map parameters from some older version to the version 2.0 specification. 

Many new parameter changes will require an entry here to allow @soroushojagh 's parameter database to map properly. If we choose to support @soroushojagh 's parameter database, which I think we should, we need to do some work here once we figure out the specification @soroushojagh  has used. This mapper will need to be maintained from now until we collectively choose to not support the older specification.

Comment (soroushojagh) [reply LDAR_Sim/model_code/input_mapper_v1.py:55]: Thanks @tbarchyn, 

For version 2.0, if changes are about values of different parameters, it can be easily handled for the Web API DB. 

Example: 

Changes form `'LPR': 0.0065 `  to `'LPR': 0.0085 `

In another scenario, if some new parameters add to the old version of input parameters, again they can be easily added to the DB. 

**My concern:** some structural changes happen in the input parameters. For example, 

`'MDL': [0.01275, 2.78e-6],` 

Changes to: 

`'Parameters1': 0.01275`

`'Parameter2': , 2.78e-6`

Then, it should be time-consuming for me on the Web API side. 

In conclusion, I don't think I need a mapper for changing values or adding new parameters from V1 to V2. But, for structural changes, I definitely need a mapper or re-design of the DB.

Comment (tbarchyn) [reply LDAR_Sim/model_code/input_mapper_v1.py:55]: There have been structural changes to the parameters and will continue to be new changes since when you built the input parameter database.

What branch of LDAR-Sim are you using for the web app right now??

Yes, an input mapper is required. It is already required for v1 files (refer to this file we are commenting on),

What I'm hoping you can provide is a branch that defines the structure of the parameters you have built the database with so I can build at least the start of the mapper. 

If possible, could you provide a branch that defines the version of LDAR sim that works with the web app.

Thanks.

Comment (soroushojagh) [reply LDAR_Sim/model_code/input_mapper_v1.py:55]: I designed the DB based on input parameters in IMTS_Private: `/master/LDAR_Sim/inputs_template`.

Comment (tbarchyn) [reply LDAR_Sim/model_code/input_mapper_v1.py:55]: OK, your branch is:

https://github.com/IM3S/LDAR_Sim/tree/Soroush_input_database

All changes to input parameters from that point forward are going to need to map from that branch to whatever the latest version is. Thanks. 

This is a priority as there are PRs that are going to change the parameters and structure of those parameters and we need to capture the structure and defaults you are using BEFORE those changes are merged - otherwise you will have a lot of unnecessary work to do!

**Comment by tbarchyn** [comment]
Decisions made:
1. Rebase following merge of method module PR.
2. Define subset of parameters for Soroush from the v2 parameter set for use within the web app - define the specification closely to limit reverse compatibility burden.
3. No reverse compatibility supported to v1

**Comment by tbarchyn** [comment]
See #90 - this one can be closed eventually as #90 replaces it.
